### ISSUE TITLE
[Question] Arguments passed to SingleObjective.eval() not to be optimized upon

### ISSUE THREAD

**Issue by alexmaraval** [SheffieldML/GPyOpt#318]
In a simple setting, I want to optimize an `objective` function like below.
```
import GPyOpt

def objective(inp):
    return inp[0][0] ** 2

bounds = [{'name': 'x', 'type': 'continuous', 'domain': (-4, 4)}]

bo = GPyOpt.methods.BayesianOptimization(objective, domain=bounds)
```
But what if I would like my `objective` function to depend on other arguments ? I found that (forgive the inelegant method) adding inputs to the `objective` input array and setting their bounds fixed works.
```
def objective(inp):
    x, z = inp[0][0], inp[0][1]
    return x ** 2 + z

bounds = [{'name': 'x', 'type': 'continuous', 'domain': (-4, 4)},
          {'name': 'z', 'type': 'continuous', 'domain': (-1, -1)}]

bo = GPyOpt.methods.BayesianOptimization(objective, domain=bounds)
```
Maybe there is a more elegant way to do that using equality constraints but that does the trick I guess.

But now, what if I would like my `objective` to have keyword arguments like so ?
```
def objective(inp, **kwargs):
    z = kwargs['z_value']
    return inp[0][0] ** 2 + z
```
In this case, how do I pass a `**kwargs` dictionary to the `SingleObjective` class so that the `eval()` method can use them ? This would be useful for example if the `objective` is a function that depends on some configuration.

I've looked in the `GPyOpt` repo and at examples online and all I could find is people declaring all the configuration in the same scope as the `objective` so that there is no need to pass `**kwargs`. But that can become very messy and it would be nice to have something like a user-defined-`**kwargs`-manager or something similar.

For example, it would be useful to be able to pass such `**kwargs` to the `objective`, maybe through the `BayesianOptimization` constructor like so
```
def objective(x, **kwargs):
    z = kwargs['z_value']
    return x ** 2 + z

bounds = [{'name': 'x', 'type': 'continuous', 'domain': (-4, 4)}]
obj_kw = {'z_value': -1}
bo = GPyOpt.methods.BayesianOptimization(objective, domain=bounds, objective_kwargs=obj_kw)
```

In general, I guess my question relates to what can we pass to the `objective` ? Is there a way to pass other arguments than variables in a 2d array ? Is there a way to pass arguments that the optimization can ignore but that are necessary to calculate the value of the `objective` ?

A more interesting example would be tuning the hyper-parameters of a network. It could be useful to pass the network, the loss and the optimizer as keyword arguments to the `objective` and have the hyper-parameters (e.g. learning rate(s)) as the arguments the objective will be optimized upon.

If there is already a way to do that I'm not aware of it and I'm sorry for the useless issue.

**Comment by apaleyes** [comment]
Check out context: https://nbviewer.jupyter.org/github/SheffieldML/GPyOpt/blob/master/manual/GPyOpt_context.ipynb

Not sure if it's a 100% match, but it seems pretty close to what you are looking for.

**Comment by alexmaraval** [comment]
Hi @apaleyes and thanks for the reply.
I don't think this is what I want. Correct me if I'm wrong but what you pointed me towards is a way to define a `mixed_domain` and then use a `context` to fix some of the variables of that `mixed_domain`.

Whereas what I was looking for is really to be able to pass anything to the objective function, like `*args` and `**kwargs`.

So I came up with this below, which essentially is a wrapper to call the objective function with some pre-set `*args` and `**kwargs`. I create this class `CustomSingleObjective` that inherits everything from class `SingleObjective` from `GPyOpt.core.task.objective`. The idea is to use a decorator `ObjectiveDecorator` to "pass" arguments to the constructor of `CustomSingleObjective` so that it itself passes the decorated `custom_func` to the superclass `SingleObjective` who whill treat this `custom_func` normally.

```
from GPyOpt.core.task.objective import SingleObjective


class ObjectiveDecorator(object):
    def __init__(self, add_args, add_kwargs):
        self.add_args = add_args
        self.add_kwargs = add_kwargs

    def __call__(self, func, *args, **kwargs):
        def new_func(*fargs, **fkwargs):
            new_args = fargs + self.add_args
            new_kwargs = {**fkwargs, **self.add_kwargs}
            return func(*new_args, **new_kwargs)
        return new_func


class CustomSingleObjective(SingleObjective):
    def __init__(self, func, *args, 
                 num_cores=1, objective_name='no_name', batch_type='synchronous', 
                 space=None, **kwargs):
        self.args = args
        self.kwargs = kwargs
        self.func = func

        @ObjectiveDecorator(args, kwargs)
        def custom_func(v, *func_args, **func_kwargs):
            return func(v, *func_args, **func_kwargs)

        super(CustomSingleObjective, self).__init__(
             custom_func, num_cores, objective_name, batch_type, space)
```
As an example I create a `toy_objective` function that takes one 2d numpy array as the normal objective of `SingleObjective` would. But I also allow it to take `*args` and `**kwargs`.
```
def toy_objective(v, *args, **kwargs):
    return v[0] ** 2 + kwargs.get('y', 0) + sum(args)
```
Now I'm able to construct an instance of `CustomSingleObjective` that would take into account `*args` and `**kwargs` and use them in a call to `evaluate()` method.
```
co = CustomSingleObjective(toy_objective, 6, -1, y=-1)
so = SingleObjective(toy_objective)

x = np.atleast_2d(np.array([2]))
a, b = so.func(x), so.evaluate(x)
c, d = co.func(x), co.evaluate(x)
print(a, b)
print(c, d)
```
which results in

> a = array([4.])
> b = (array([[4.]]), [8.392333984375e-05])
> c = array([8])
> d = (array([[8.]]), [0.00014019012451171875])

So the arguments passed to the `CustomSingleObjective` constructor are indeed used in the computation of `toy_objective` when calling the `evaluate()` method from the mother class `SingleObjective`.

Although this is nice and all, I'm not 100% confident in my decorator skills so there may be an easier way to do the same thing. Also I haven't yet tested this on my NeuralNet tuning case yet but I have hope this could help.

Let me know what you think and if you find it useful I think this could be easily added to `GPyOpt/core/task/objective.py` to have an intuitive way to pass all different kinds of arguments to the objective function that have nothing to do with the optimization itself.

**Comment by apaleyes** [comment]
Thanks for following up on this!

The way you approached this looks sensible, since there is nothing built into GPyOpt to support this use case.

Given that this is the first time I come across such a request, we shouldn't rush to include it into the core library just yet. I made sure to add it to the [FAQ](https://github.com/SheffieldML/GPyOpt/wiki/FAQ), so it is discoverable both here and there. If we see that more people express interest in this feature, we may end up merging it in.

Again, thanks a lot for sharing your solution.

**Comment by alexmaraval** [comment]
Makes sense, thanks for putting it in the FAQ (hope it will be useful) and thanks for the feedback
:)
